src/main/java/dev/blaauwendraad/masker/json/ValueMasker.java

```diff
@@ -0,0 +1,168 @@
+        );
+    }
+
+    /**
+     * Does not mask a target value.
+     *
+     * @see KeyMaskingConfig.Builder#disableBooleanMasking()
+     * @see KeyMaskingConfig.Builder#disableNumberMasking()
+     */
+    static ValueMasker noop() {
+        return withDescription("<no masking>", context -> {
+        });
+    }
+
+    /**
+     * Masks a target string value (containing an email) while keeping some amount of the prefix characters and ability
+     * to keep the domain unmasked.
+     * For example:
+     * {@literal "maskMe": "agavlyukovskiy@gmail.com" -> "maskMe": "***@gmail.com"}
+     * {@literal "maskMe": "agavlyukovskiy@gmail.com" -> "maskMe": "ag***"}
+     * {@literal "maskMe": "agavlyukovskiy@gmail.com" -> "maskMe": "ag***@gmail.com"}
+     * {@literal "maskMe": "agavlyukovskiy@gmail.com" -> "maskMe": "ag***iy@gmail.com"}
+     *
+     * @param keepPrefixLength amount of prefix characters to keep unmasked
+     * @param keepDomain if true - the email domain will remain unmasked
+     * @param mask the static mask applied to the rest of the value
+     */
+    static ValueMasker maskEmail(int keepPrefixLength, int keepSuffixLength, boolean keepDomain, String mask) {
+        byte[] replacementBytes = mask.getBytes(StandardCharsets.UTF_8);
+        String description = "email, keep prefix: %s, keep suffix: %s, keep domain: %s"
+                .formatted(keepPrefixLength, keepSuffixLength, keepDomain);
+        return withDescription(description, context -> {
+            int prefixLength = keepPrefixLength + 1; // add opening quote
+            int suffixLength = keepSuffixLength + 1; // keep closing quote
+            if (keepDomain) {
+                for (int i = 0; i < context.valueLength(); i++) {
+                    if (context.getByte(i) == '@') {
+                        // include domain in the suffix
+                        suffixLength = context.valueLength() - i + keepSuffixLength;
+                        break;
+                    }
+                }
+            }
+            int maskLength = context.valueLength() - prefixLength - suffixLength;
+            if (maskLength > 0) {
+                context.replaceValue(prefixLength, maskLength, replacementBytes, 1);
+            }
+        });
+    }
+
+    /**
+     * Masks a target value with a supplied {@link Function}. The target value is passed into the function as a string,
+     * regardless of the type (string, numeric or a boolean), however any non-null returned value from the
+     * function will always be a JSON string (with quotes added automatically).
+     *
+     * <p>Note: usually {@link ValueMasker} operates on a byte level without parsing JSON values into intermediate
+     * objects, however this implementation will have to allocate a {@link String} before passing it into the function
+     * and then turn it back into a byte array for replacement.
+     */
+    static ValueMasker maskWithStringFunction(Function<String, String> masker) {
```

Comment:
Based purely on API client's perspective, it would be more intuitive if I can just pass a lambda `maskNumber((value) -> "888"`, then for presets I do `maskString(ValueMasker::maskEmail)`. Then, on the builder layer we can decorate these lambdas with DescriptiveMakser. Also, `MaskingStrategy` would look more intuitive than `ValueMakser`.

Reply:
But I think the current API is also intuitive enough.

Reply:
> it would be more intuitive if I can just pass a lambda maskNumber((value) -> "888"

I think that's a nice idea, we could add another builder method like this
```java
        public Builder maskStringsWith(Function<String, String> masker) {
            maskStringsWith(ValueMasker.maskWithStringFunction(masker));
            return this;
        }
```
but the only problem is that currently `Function<String, String>` behaves more like "string representation of the value", which also applied to numeric types. We could add more of these, for example
```java
public Builder maskNumbersWith(Function<Number, Number> value)
public Builder maskBooleansWith(Function<Boolean, Boolean> value)
```
but then we quickly run into problem of impossibility to convert type during masking, there can't be two methods like this:
```java
public Builder maskNumbersWith(Function<Number, Number> value)
public Builder maskNumbersWith(Function<Number, String> value)
```

Reply:
The problem still holds though regardless of whether we pass a lambda directly or decorate it through ValueMasker. It's just the first way that is more intuitive. 
As for the problem itself, I think it is unlikely that someone would need to partially mask numbers without converting the value to strings. Same for booleans.

Reply:
> The problem still holds though regardless of whether we pass a lambda directly or decorate it through ValueMasker.

This is not entirely true - inside `ValueMasker` we actually differentiate by the type of the function, so in addition to `ValueMasker.maskWithStringFunction(Function<String, String>)` we could add `ValueMasker.maskWithNumberFunction(Function<Number, Number>)` and `ValueMasker.maskWithNumberStringFunction(Function<Number, String>)`

> As for the problem itself, I think it is unlikely that someone would need to partially mask numbers without converting the value to strings. Same for booleans.

WIth partial masking yes, but this is more about generic function for masking, as a user I'd rather expect `Function<Number, Number>` to be the default (though working with `Number` and not concrete implementation is also not very friendly)